### PR TITLE
Support forwarding container logs to syslog

### PR DESCRIPTION
**Pull request by maximkulkin**
Docker has feature to view container logs (output to stdout and stderr),
but it is not very convenient. This patch adds an option to enable
forwarding of log messages to syslog server.

To enable syslog forwarding, add "--syslog" to docker's "run" command's
arguments. It will forward messages to local (host's) syslog server. To
forward to remote syslog, use "--syslog-url=udp://server:port"
parameter.

Other options allow configuring syslog parameters:
- --syslog-facility
- --syslog-severity
- --syslog-tag

### CODE REVIEW

**Comment by maximkulkin** [comment]
@mhart as of separating STDOUT and STDERR, I would prefer
1) hear more demand for that feature from others;
2) implemented it as a separate pull request once this pull request has landed (passing though a separate review/discussion).

**Comment by maximkulkin** [comment]
@SvenDowideit please point me to the sentence that you think can be misinterpreted.

**Comment by jfrazelle** [comment]
I do feel like this is a stop gap for actual logging drivers and maybe we should consider a way of turning off logging for containers all together first. For example irssi or mutt containers.

**Comment by maximkulkin** [comment]
The larger your feature the more haters you get

**Comment by erikh** [comment]
+1

File: daemon/daemon.go [hunk 1]
```diff
@@ -299,6 +302,64 @@ func (daemon *Daemon) LogToDisk(src *broadcastwriter.BroadcastWriter, dst, strea
...
+	return w.next.Close()
+}
+
+func (daemon *Daemon) LogToSyslog(src *broadcastwriter.BroadcastWriter, syslogConfig runconfig.SyslogConfig) error {
+	var (
+		log io.WriteCloser
+		err error
+	)
+	if syslogConfig.Url != "" {
+		u, err := url.Parse(syslogConfig.Url)
```
Comment (ambakshi) [inline daemon/daemon.go]: Doesn't this redeclare 'err' within local scope and shadow the function's err checked outside?

**Comment by SvenDowideit** [comment]
@maximkulkin there is no sentence - which is the problem, you need to add more information to clarify to the reader the scope and limitations of the feature, otherwise they are left to imagine, which leads to needless bug reports.

**Comment by fredlf** [comment]
I agree with Sven that docs should explicitly state that `syslog` options don't affect `docker log` options, etc. Otherwise, docs seem adequate. There are copy editing issues that can be addressed in a follow-on PR.

**Comment by maximkulkin** [comment]
I agree on the statement that docs should be clear. I added a line that _explicitly_ states that `docker logs` functionality wasn't affected.

**Comment by maximkulkin** [comment]
@ambakshi Thanks for pointing this out. Fixed.

File: daemon/daemon.go (hunk 1 above)
Comment (maximkulkin) [inline daemon/daemon.go]: Thanks for pointing this out. Fixed.

**Comment by fredlf** [comment]
Docs LGTM We can clean up and add more if needed in a follow-on PR. Ping @SvenDowideit @jamtur01

**Comment by Kokonoe** [comment]
+1
Should be nice to use it once i'm using Logspout (https://github.com/progrium/logspout).

File: docs/sources/reference/commandline/cli.md [hunk 2]
```diff
@@ -1674,6 +1682,19 @@ An example of a file passed with `--env-file`
 This will create and run a new container with the container name being
 `console`.
 
+    $ sudo docker run --syslog --syslog-tag MONGODB mongo:latest
+
+The `--syslog` flag will make Docker forward logs to the host system's log server.
```
Comment (SvenDowideit) [inline docs/sources/reference/commandline/cli.md]: can we just change this to `The `--syslog` flag will make Docker forward logs to the system's log server in addition to sending it to Docker's internal log store.`

please?

File: docs/man/docker-run.1.md [hunk 3]
```diff
@@ -265,6 +270,21 @@ outside of a container on the host.
 **--sig-proxy**=*true*|*false*
    Proxy received signals to the process (non-TTY mode only). SIGCHLD, SIGSTOP, and SIGKILL are not proxied. The default is *true*.
 
+**--syslog**=*true*|*false*
+   When set to true Docker will send all STDOUT and STDERR output to the syslog server. The default is *false*.
```
Comment (SvenDowideit) [inline docs/man/docker-run.1.md]: please add 'in addition to sending it to Docker's internal log store`

File: runconfig/parse.go [hunk 4]
```diff
@@ -62,6 +64,12 @@ func Parse(cmd *flag.FlagSet, args []string) (*Config, *HostConfig, *flag.FlagSe
 		flMacAddress      = cmd.String([]string{"-mac-address"}, "", "Container MAC address (e.g. 92:d0:c6:0a:29:33)")
 		flIpcMode         = cmd.String([]string{"-ipc"}, "", "Default is to create a private IPC namespace (POSIX SysV IPC) for the container\n'container:<name|id>': reuses another container shared memory, semaphores and message queues\n'host': use the host shared memory,semaphores and message queues inside the container.  Note: the host mode gives the container full access to local shared memory and is therefore considered insecure.")
 		flRestartPolicy   = cmd.String([]string{"-restart"}, "", "Restart policy to apply when a container exits (no, on-failure[:max-retry], always)")
+
+		flSyslog         = cmd.Bool([]string{"-syslog"}, false, "Send logs to Syslog server")
```
Comment (SvenDowideit) [inline runconfig/parse.go]: perhaps change this to `Also send logs to Syslog server`

**Comment by SvenDowideit** [comment]
LGTM, after addressing my wording changes - I really don't want anyone to missinterpret what this does.

**Comment by maximkulkin** [comment]
@SvenDowideit done

**Comment by wuub** [comment]
@SvenDowideit any ETA on merging this? no reliable way to combine docker container logs with the rest of our system is so painful for us.

**Comment by SvenDowideit** [comment]
@wuub its not up to me - it needs a design review and a code review - @crosbymichael @tiborvass @jfrazelle ?

**Comment by maximkulkin** [comment]
Rebased to latest master

**Comment by tiborvass** [comment]
I'm sorry to be the annoying guy this time but I'm -1 on this if only for flag pollution.

I agree with @cpuguy83 here that this is better done in a logging driver.
You are right that the logging driver has not happened yet and it's been some time: our bad.

What I suggest is that we make a special effort to include proper logging drivers in 1.6. Your work in this PR can help us make sure the driver interfaces work well, and your code could be reused.

That was my personal opinion :)

**Comment by cnf** [comment]
After looking at the source of this, it just calls go-lang `log/syslog`. The problem is that `log/syslog` does NOT support [RFC3164 section 4.1.2](https://tools.ietf.org/html/rfc3164#section-4.1.2). Specifically, the timestamp is hardcoded to RFC3339 [right here](https://github.com/golang/go/blob/428afae027548e079335d25cacb70082a0ef8d9f/src/log/syslog/syslog.go#L298).

RFC3164 (BSD) style syslog is still used a lot, especially on legacy systems, or commercial systems, so not implementing it would have my -1 (but a definite +1 to syslog in general)

**Comment by maximkulkin** [comment]
Definitely not going to invest in neither implementing support for all bizare syslog implementations, nor to fix all software in the internets.

**Comment by cnf** [comment]
I am sorry @maximkulkin, did you just call the official syslog RFC a "bizarre syslog implementation"? I am not suggesting to implement some exotic format, just the actual RFC specification. In my opinion, standards are important, and we should try to adhere to them.

**Comment by maximkulkin** [comment]
Look, there is a syslog client in Go *standard library*. Now you come and tell me that *their* implementation is flawed and *I* need to reinvent the wheel and make my own implementation that does not have the formentioned flaw. My opinion is that you should complain to Go standard library issue tracker about non-supported format and not compliance to RFC.

This happens all the time when you try to contribute and then somebody comes and asks to fix all problems with all software in the internets. That's not going to happen. If you need it, go and fix it yourself.

**Comment by cnf** [comment]
I am actually in the process of getting log/syslog fixed, because it doesn't follow the standards.

The go people have been quite supportive on this, so far, but it won't be fixed by tomorrow.

So I guess -1 from me, until log/syslog can be updated...

**Comment by maximkulkin** [comment]
So, your logic is like "If it doesn't work for me, let the other people suffer the same". OK. FINE

**Comment by dangra** [comment]
> So I guess -1 from me, until log/syslog can be updated...

You must be kidding if you prefer to block a feature based on an external and orthogonal issue. When it got fixed in go-lang it will be ready for you to submit a fix here if needed.

**Comment by dangra** [comment]
> I am actually in the process of getting log/syslog fixed, because it doesn't follow the standards.
> The go people have been quite supportive on this, so far, but it won't be fixed by tomorrow.

@cnf at least link to the discussion/issue/ticket/whatever in go-lang side so others reading your comments can get an idea of the status of this *so-needed* feature of yours.

**Comment by cnf** [comment]
It has been on #go-lang so far. 

I'm sorry, but I do not understand the attitude against trying to get docker standards compliant. What I am asking is not unreasonable.

Either way, I am going to FOSDEM this weekend, but coming week I am actually looking at this issue for `log/syslog` for several other projects. I need to read up on how to actually get PR accepted into stdlib. Can everyone agree to hold out on this bit for a while so I can see of this is feasible for me?

**Comment by maximkulkin** [comment]
There is no "attitude", it IS standard compliant: there is a [The syslog protocol RFC](https://tools.ietf.org/html/rfc5424) which mentions correct timestamp format (the one that is used in Go's syslog library).

**Comment by dangra** [comment]
> I'm sorry, but I do not understand the attitude against trying to get docker standards compliant. What I am asking is not unreasonable.

Nobody stops you and I am actually in favor of supporting standards, but makes no sense to put on hold features now that will benefit of future stdlib improvements. Once stdlib supports RFC3164 timestamp you will see them in docker, it may require some changes but it won't be as hard as getting a complete new feature merged into docker codebase.

That's the hard part, getting the code and functionality in acceptable minimal form to be merged. You're undermining it based on a feature that should be somewhere else.

Still, rest in peace, you still have time as the maintainers are unlikely to merge this PR because they prefer to do it *the right way* even if it takes longer. I'm sorry because I really feel the need for something like what @maximkulkin  implemented, it will allow me to drop the log-courier/logstash-forwarder setup. I also wished this PR added a way to prevent docker from logging lot local files but I understand why it can't happen now :)

**Comment by erikh** [comment]
Hi, we prefer a proposal in documentation form that has been approved before we accept code changes.

Bugfixes are exempt from this, but this is definitely a new feature.

-Erik

**Comment by maximkulkin** [comment]
LOL, WAT? Now they go bureaucratic.

**Comment by erikh** [comment]
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#design-and-cleanup-proposals <https://github.com/docker/docker/blob/master/CONTRIBUTING.md#design-and-cleanup-proposals> highlights our contribution process.

-Erik

**Comment by maximkulkin** [comment]
OMG. I'm not doing this b*shit. I waste my personal time trying to make software to be a better place and you just go douchebags.

**Comment by erikh** [comment]
Ok, I'm going to close this PR then. Sorry we couldn't find a way for you to contribute.

**Comment by maximkulkin** [comment]
Oh, thank you. Now we all know what is like to contribute to Docker.

**Comment by crosbymichael** [comment]
@maximkulkin sorry that this has been a frustrating experience for you - this discussion could have been handled better, and we appreciate the time and effort you put into this.

It’s very valuable to have members of the community commenting and providing information that is relevant to changes because it helps to have people with various use cases and experiences provide feedback. However, if you’re unsure who has an authoritative voice on the final changes you can refer to this [documentation](https://github.com/docker/docker/blob/master/MAINTAINERS) with maintainers' names and email addresses.

Logging is still very important and a big pain for many users. We currently have a proposal for improved logging support in docker #7195 and your work on syslog is definitely something that we want to support as it is so common. It's unfortunate that we have not been able to focus on this just yet, but seeing more issues, PRs, and proposals helps to remind everyone in the community that this is still a huge issue.

The current contribution process located [here](https://github.com/docker/docker/blob/master/CONTRIBUTING.md#design-and-cleanup-proposals) may not be perfect and it may seem too process-heavy to you but following it makes it more likely that your work will end up being approved.

Again, sorry that this has been frustrating and if you wish to join the conservation around improved logging please jump in on #7195. Just remember to please respond politely and be respectful to other members of the community as any feedback and comments are meant to be helpful.

**Pull request closed without merging**